armi/reactor/composites.py

```diff
@@ -771,66 +778,80 @@ def _deriveUndefinedVolume(self):
+
+        otherCompAreas = sum([c.getArea() for c in otherComps])
+        otherCompVolumes = sum([c.getVolume() for c in otherComps])
+
+        remainingArea = self.getMaxArea() - otherCompAreas
+        remainingVolume = self.getMaxVolume() - otherCompVolumes
+
+        geometryError = False
+        # Check for negative area
+        if remainingArea < 0:
+            geometryError = True
+            componentAreas = "\n".join(
+            [
+                f"{c:40s}, {c.getArea():.5e} cm^2"
+                for c in otherComps
+            ])
+            msg = (f"The component areas in {self} exceed the maximum "
+                   f"allowable area based on the geometry. Check that the "
+                   f"geometry is defined correctly.\n"
+                   f"Maximum allowable area: {self.getMaxArea()} cm^2\n"
+                   f"Area of all defined components: {otherCompAreas} cm^2\n"
+                   f"Breakdown: {componentAreas}")
+            runLog.error(msg)
+
+        # Check for negative volume
+        if remainingVolume < 0:
+            geometryError = True
+            componentVols = "\n".join(
+            [
+                f"{c:40s}, {c.getVolume():.5e} cc"
+                for c in otherComps
+            ])
+            msg = (f"The component volumes in {self} exceed the maximum "
+                   f"allowable volume based on the geometry. Check that the "
+                   f"geometry is defined correctly.\n"
+                   f"Maximum allowable volume: {self.getMaxVolume()} cc\n"
+                   f"Volume of all defined components: {otherCompVolumes} cc\n"
+                   f"Breakdown: {componentVols}")
+            runLog.error(msg)
+
+        if geometryError:
+            raise ValueError(f"Negative area/volume errors occurred for {self}. "
+                             "Check log for errors.")
+
+        if len(derivedShapeComps) == 1:
+            c = derivedShapeComps[0]
+
+            # A ZeroDivisonError can occur when the height of the component's
+            # parent is zero, thus the volume may be set within `setVolume`
+            # (using the derived volume calculated here), but the component's
+            # area cannot be determined by directly using this volume. In this
+            # case, the area can be computed by subtracting the area of all
+            # other components from the parent's maximum area.
+            try:
+                c.setVolume(remainingVolume)
+            except ValueError:
+                c.setArea(remainingArea)
+
+        elif derivedShapeComps > 1:
+            msg = (f"The number of `DerivedShape` children exceeds 1 in {self} "
+                   f"and it is not possible to set the volume/area of these children.\n"
+                   f"Remove the multiple instances of these components to fully define the "
+                   f"geometry.\n"
+                   f"Derived shapes: {derivedShapeComps}")
```

Comment:
I think `"More than 1 DerivedShapes ({derivedShapeComps}) in {self} is not allowed"` is basically sufficient here.